README.md
#### Note: This repository is inherited from the [old Tuya Github repository](https://github.com/TuyaInc/tuyasmart_home_ios_sdk), which will be deprecated soon. Please use this repository for Tuya SDK development instead. For changing the existing remote repository URL, please check this [tutorial](https://docs.github.com/en/free-pro-team@latest/github/using-git/changing-a-remotes-url).

# Tuya Smart iOS SDK

[English](README.md)|[中文版](README-zh.md)

---

## Features Overview

Tuya Smart iOS Home SDK is the iOS 9.0 and above version provided by Tuya for the field of smart home. iOS developers can quickly develop app functions based on SDK, realize the activation of intelligent hardware, hardware control, firmware upgrades, timed tasks, intelligent scenarios and other operations.

The SDK includes the following features:

- Account system (phone number, email registration, login, password reset and other general account functions)
- Home system (home management, room management, home sharing and other functions)
- Hardware functions (network configuration, control, status reporting, timed tasks, groups, firmware upgrades, sharing)

## Fast Integration

### Using CocoaPods integration (version 9.0 or above is supported)

Add the following content in file `Podfile`:

```ruby
platform :ios, '9.0'

target 'your_target_name' do

      pod "TuyaSmartHomeKit"

end
```

Execute command `pod update` in the project's root directory to begin integration.

For the instructions of CocoaPods, please refer to: [CocoaPods Guides](https://guides.cocoapods.org/)

## Initializing SDK

1. Open project setting, `Target => General`, edit `Bundle Identifier` to the value from Tuya develop center.
2. Import security image to the project and rename as `t_s.bmp`, then add it into `Project Setting => Target => Build Phases => Copy Bundle Resources`.
3. Add the following to the project file `PrefixHeader.pch`：

```objective-c
#import <TuyaSmartHomeKit/TuyaSmartKit.h>
```

Open file `AppDelegate.m`，and use the `App ID` and `App Secret` obtained from the development platform in the `[AppDelegate application:didFinishLaunchingWithOptions:]`method to initialize SDK:

```objective-c
[[TuyaSmartSDK sharedInstance] startWithAppKey:<#your_app_key#> secretKey:<#your_secret_key#>];
```

Now all the preparatory work has been completed. You can set out to develop your application.

## Doc

Refer to details: [Tuya Smart Doc - iOS SDK](https://tuyainc.github.io/tuyasmart_home_ios_sdk_doc/en/)

## ChangeLog

[Change Log](https://developer.tuya.com/en/docs/app-development/ios-app-sdk/changelog?id=Ka5vc933r3qja)


## Support

You can get support from Tuya with the following methods:

Tuya Smart Help Center: https://support.tuya.com/en/help

Technical Support Council: https://service.console.tuya.com

## License

This Tuya Home iOS SDK Sample is licensed under the MIT License.


CHANGELOG.md
# Tuya Smart iOS Home SDK Release Notes

## v3.24.0

### Release Date

2020.02.05

### Highlights

- Provided low-code generic panel for controlling smart devices, namely the **TuyaSmartDefaultPanelKit** library.
- Optimized third-party library compatibility. The dependency library is changed to the dynamic library (OpenSSL and MQTTClient library).
- Optimized OTA for low power devices.
- Supported NB-IoT devices sharing.

<div>
        &copy; 2014 - 2021 Tuya Inc. All Rights Reserved.
</div>

## v3.22.0

### Release Date

2020.12.25

### Highlights

- Add binding email feature
- Add Homekit device activation feature
- Support infrared RF devices
- Add big data channel entrance
- Add standard electrician product category group

### Fixed Bugs

- Optimize the problem of slow subscription when there are too many user devices
- Optimize the problem of sharing Bluetooth devices without opening

<div>
        &copy; 2014 - 2021 Tuya Inc. All Rights Reserved.
</div>


## v3.20.0

### Release Date

2020.10.30

### Highlights

- New guest user login mode
- Support new lightning active device
- Bugs fixed

<div>
        &copy; 2014 - 2021 Tuya Inc. All Rights Reserved.
</div>


## v3.19.0

### Release Date

2020.09.21

### Highlights

- New timing interface function
- The room supports changing custom pictures
- Added device dps time for device attributes, which can record the time corresponding to the nearest change of dp
- Bugs fixed

<div>
        &copy; 2014 - 2021 Tuya Inc. All Rights Reserved.
</div>


## v3.17.6

### Release Date

2020.07.09

### Highlights

- Smart scene of timing conditions related to sunrise and sunset
- Bugs fixed

<div>
        &copy; 2014 - 2021 Tuya Inc. All Rights Reserved.
</div>


## v3.17.0

### Release Date

2020.04.27

### Highlights

- Smart scene support for geofence condition trigger
- Firmware upgrade status prompt
- QR code login
- Mesh support fast active
- Support Wi-Fi sub-device group
- Bugs fixed

<div>
        &copy; 2014 - 2021 Tuya Inc. All Rights Reserved.
</div>


## v3.15.0

### Release Date

2020.04.11

### Highlights

- Group support set custom icon
- Bugs fixed

<div>
        &copy; 2014 - 2021 Tuya Inc. All Rights Reserved.
</div>


## v3.14.0

### Release Date

2019.12.21

### Highlights

- MQTTClient Use 0.15.2
- HTTP Request and Response Data Encryption
- Manager More Family Member Role
- Add Scene Cache
- Add Scene Conditions and Actions Class
- Support for Login with Apple
- Bugs Fixed

<div>
        &copy; 2014 - 2021 Tuya Inc. All Rights Reserved.
</div>


## v3.13.0

### Release Date

2019.12.06

### Highlights

- Support DP Code Control
- Support More Family Member Role

<div>
        &copy; 2014 - 2021 Tuya Inc. All Rights Reserved.
</div>


## v3.12.0

### Release Date

2019.09.21

### Highlights

- Device Support Update Head Image
- Support Wi-Fi and Bluetooth Dual-mode Devices
- Support Sub Device OTA
- Firmware Upgrade to Add Callbacks in Upgrading
- AP Distribution Network Optimization
- Timing Support Annotation and Push Function
- Support SigMesh Devices
- Bugs Fixed

<div>
        &copy; 2014 - 2021 Tuya Inc. All Rights Reserved.
</div>


## v2.12.46

### Release Date

2019.07.22

### Highlights

- Push Notification Was Classified
- Bugs Fixed

<div>
        &copy; 2014 - 2021 Tuya Inc. All Rights Reserved.
</div>


## v2.10.97

### Release Date

2019.06.20

### Highlights

- Support Cocoapods (>=1.6.0)

<div>
        &copy; 2014 - 2021 Tuya Inc. All Rights Reserved.
</div>


## v2.10.96

### Release Date

2019.05.11

### Highlights

- Added Device Network Configuration Without Password
- Added Push Notification to Open or Close
- Add Development Environment Push Support
- Bugs Fixed

<div>
        &copy; 2014 - 2021 Tuya Inc. All Rights Reserved.
</div>


## v2.8.44

### Release Date

2019.03.28

### Highlights

- Fixed an Online Problem with Wi-Fi Subdevices
- Bugs Fixed

<div>
        &copy; 2014 - 2021 Tuya Inc. All Rights Reserved.
</div>


## v2.8.43

### Release Date

2019.03.15

### Highlights

- New Encryption: Appkey, AppSecret, iOS Security Images, BundleId for Application on Tuya Developer Platform

<div>
        &copy; 2014 - 2021 Tuya Inc. All Rights Reserved.
</div>

TuyaSmartDemo/Classes/AddDevice/Controller/TYDemoAddBLEMeshViewController.m
//
//  TYAddBLEMeshViewController.m
//  TuyaSmartHomeKit_Example
//
//  Created by milong on 2020/7/21.
//  Copyright © 2020 xuchengcheng. All rights reserved.
//

#import "TYDemoAddBLEMeshViewController.h"
#import "TYDemoAddDeviceUtils.h"
#import "TYDemoConfiguration.h"
#import "TYDemoBleDeviceListViewController.h"
#import "TPDemoProgressUtils.h"

@interface TYDemoAddBLEMeshViewController () <TuyaSmartSIGMeshManagerDelegate>

@property (nonatomic, strong) UILabel     *productIdLabel;
@property (nonatomic, strong) UITextField *productIdTextField;
@property (nonatomic, strong) UILabel     *consoleLabel;
@property (nonatomic, strong) UITextView  *consoleTextView;
@property (nonatomic, strong) UIButton    *scanButton;
@property (nonatomic, strong) UIButton    *stopScanButton;
@property (nonatomic, strong) UIAlertController *actionSheet;
//扫描到的ble mesh设备
@property (nonatomic, strong) NSMutableArray<TuyaSmartSIGMeshDiscoverDeviceInfo *> *meshModelArray;
//配网成功设备node Id 集合
@property (nonatomic, strong) NSMutableArray *nodeIdArray;
//配网成功设备 device id集合
@property (nonatomic, strong) NSMutableArray *devIdArray;

@property (nonatomic, strong) TYAlertMessgeView *alertMessageView;
@end

@implementation TYDemoAddBLEMeshViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    [self initView];
}

- (void)initView {
    self.view.backgroundColor = [UIColor whiteColor];
    self.topBarView.leftItem = self.leftBackItem;
    self.centerTitleItem.title = @"Add BLE Mesh";
    self.topBarView.centerItem = self.centerTitleItem;
    
    CGFloat currentY = self.topBarView.height;
    currentY += 10;
    CGFloat labelWidth = 75;
    CGFloat textFieldWidth = APP_SCREEN_WIDTH - labelWidth - 30;
    CGFloat labelHeight = 44;
    
    //init product id
    self.productIdLabel.frame = CGRectMake(10, currentY, labelWidth, labelHeight);
    [self.view addSubview:self.productIdLabel];
    self.productIdTextField.frame = CGRectMake(labelWidth + 20, currentY, textFieldWidth, labelHeight);
    [self.view addSubview:self.productIdTextField];
    currentY += labelHeight;
    
    //init console
    currentY += 10;
    self.consoleLabel.frame = CGRectMake(10, currentY, labelWidth, labelHeight);
    [self.view addSubview:self.consoleLabel];
    currentY += labelHeight;
    self.consoleTextView.frame = CGRectMake(10, currentY, APP_SCREEN_WIDTH - 20, 220);
    [self.view addSubview:self.consoleTextView];
    currentY += self.consoleTextView.frame.size.height;
    
    //init button
    currentY += 10;
    self.scanButton.frame = CGRectMake(10, currentY, APP_SCREEN_WIDTH - 20, labelHeight);
    [self.view addSubview:self.scanButton];
    currentY += labelHeight;
    
    currentY += 10;
    self.stopScanButton.frame = CGRectMake(10, currentY, APP_SCREEN_WIDTH - 20, labelHeight);
    [self.view addSubview:self.stopScanButton];
    
}

- (void)appendConsoleLog:(NSString *)logString {
    if (!logString) {
        logString = [NSString stringWithFormat:@"%@ : param error",NSStringFromSelector(_cmd)];
    }
    NSString *result = self.consoleTextView.text?:@"";
    result = [[result stringByAppendingString:logString] stringByAppendingString:@"\n"];
    self.consoleTextView.text = result;
    [self.consoleTextView scrollRangeToVisible:NSMakeRange(result.length, 1)];
}

/// 扫描蓝牙设备
- (void)scanMeshDevice {
    [self appendConsoleLog:@"start listening mesh devices ..."];
    [self.meshModelArray removeAllObjects];
    [self.devIdArray removeAllObjects];
    id<TYDemoDeviceListModuleProtocol> impl = [[TYDemoConfiguration sharedInstance] serviceOfProtocol:@protocol(TYDemoDeviceListModuleProtocol)];
    long long homeId = [impl currentHomeId];
    TuyaSmartBleMeshModel *sigMeshModel = [TuyaSmartHome homeWithHomeId:homeId].sigMeshModel;
    [[TuyaSmartSIGMeshManager sharedInstance] startScanWithScanType:ScanForUnprovision meshModel:sigMeshModel];
    [TuyaSmartSIGMeshManager sharedInstance].delegate = self;
}


- (void)stopScanMeshDeviceClick {
    [TuyaSmartSIGMeshManager.sharedInstance stopActiveDevice];
}

- (void)showAllMessDevice {
    WEAKSELF_AT
    if (!self.alertMessageView) {
        self.alertMessageView = [[TYAlertMessgeView alloc] initWithFrame:CGRectMake(0, APP_SCREEN_HEIGHT - 210, APP_SCREEN_WIDTH, 200)];
        [self.view addSubview:self.alertMessageView];
        [self.alertMessageView showTtile:@"扫描设备" Message:[NSString stringWithFormat:@"共扫描到 %lu mesh设备",(unsigned long)self.meshModelArray.count] completeBlock:^{
            //跳转到下个展示页面
            [weakSelf_AT meshListView];
        } cancelBlock:^{
            [weakSelf_AT.meshModelArray removeAllObjects];
            [TuyaSmartSIGMeshManager.sharedInstance stopActiveDevice];
        }];
    } else {
        [self.alertMessageView updateMessage:[NSString stringWithFormat:@"共扫描到 %lu mesh设备",(unsigned long)self.meshModelArray.count]];
    }
    
    
}

- (void)meshListView {
    WEAKSELF_AT
    TYDemoBleDeviceListViewController *deviceListVC = [[TYDemoBleDeviceListViewController alloc] init];
    deviceListVC.meshModelArray = [NSArray arrayWithArray:self.meshModelArray];
    [deviceListVC setSelectedDeviceBlock:^(NSMutableArray<TuyaSmartSIGMeshDiscoverDeviceInfo *> * _Nonnull selectedDevices) {
        [weakSelf_AT activeBleMesh:selectedDevices];
    }];
    [self.navigationController pushViewController:deviceListVC animated:YES];
}

- (void)activeBleMesh:(NSArray<TuyaSmartSIGMeshDiscoverDeviceInfo *> *)messArray {
    id<TYDemoDeviceListModuleProtocol> impl = [[TYDemoConfiguration sharedInstance] serviceOfProtocol:@protocol(TYDemoDeviceListModuleProtocol)];
    long long homeId = [impl currentHomeId];
    TuyaSmartBleMeshModel *sigMeshModel = [TuyaSmartHome homeWithHomeId:homeId].sigMeshModel;
    [[TuyaSmartSIGMeshManager sharedInstance] startActive:self.meshModelArray meshModel:sigMeshModel];
    
    [TPDemoProgressUtils showMessag:@"start activate" toView:self.view];
    
    WEAKSELF_AT
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(3 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [TPDemoProgressUtils hideHUDForView:weakSelf_AT.view animated:NO];
    });
}


#pragma mark -- TuyaSmartSIGMeshManager delegate

/// 扫描到待配网的设备
/// @param manager  mesh manager
/// @param device  待配网设备信息
- (void)sigMeshManager:(TuyaSmartSIGMeshManager *)manager didScanedDevice:(TuyaSmartSIGMeshDiscoverDeviceInfo *)device  {
    [self.meshModelArray addObject:device];
    [self appendConsoleLog:[NSString stringWithFormat:@"find device,device id is: %@",device.mac]];
    [self showAllMessDevice];
}

/**
 激活设备成功回调

 @param manager mesh manager
 @param device 设备
 @param devId 设备 Id
 @param error 激活中的错误，若发生错误，`name` 以及 `deviceId` 为空
*/

- (void)sigMeshManager:(TuyaSmartSIGMeshManager *)manager didActiveSubDevice:(nonnull TuyaSmartSIGMeshDiscoverDeviceInfo *)device devId:(nonnull NSString *)devId error:(nonnull NSError *)error {
    [self appendConsoleLog:[NSString stringWithFormat:@"activate success! mac:%@",device.mac]];
    [self.devIdArray addObject:device];
    [TPDemoProgressUtils showMessag:[NSString stringWithFormat:@"finish activate count of devices:%ld/%ld",self.devIdArray.count,self.meshModelArray.count] toView:self.view];
    
}

- (void)sigMeshManager:(TuyaSmartSIGMeshManager *)manager didFailToActiveDevice:(nonnull TuyaSmartSIGMeshDiscoverDeviceInfo *)device error:(nonnull NSError *)error {
    [self appendConsoleLog:[NSString stringWithFormat:@"activate failed! devId Id:%@",device.mac]];
}

- (void)didFinishToActiveDevList {
    
   
    [self performSelector:@selector(hideHudView) withObject:nil afterDelay:2.0];
    [self appendConsoleLog:@"activate finish"];
}

- (void)hideHudView {
     [TPDemoProgressUtils showSuccess:@"activate finish" toView:self.view];
}


#pragma mark -- setter and getter method

- (UILabel *)productIdLabel {
    if (!_productIdLabel) {
        _productIdLabel = [sharedAddDeviceUtils() keyLabel];
        _productIdLabel.text = @"product id:";
    }
    return _productIdLabel;
}

- (UITextField *)productIdTextField {
    if (!_productIdTextField) {
        _productIdTextField = [sharedAddDeviceUtils() textField];
        _productIdTextField.placeholder = @"product id cannot b entered";
        _productIdTextField.enabled = NO;
    }
    return _productIdTextField;
}

- (UILabel *)consoleLabel {
    if (!_consoleLabel) {
        _consoleLabel = [sharedAddDeviceUtils() keyLabel];
        _consoleLabel.text = @"console:";
    }
    return _consoleLabel;
}

- (UITextView *)consoleTextView {
    if (!_consoleTextView) {
        _consoleTextView = [UITextView new];
        _consoleTextView.layer.borderColor = UIColor.blackColor.CGColor;
        _consoleTextView.layer.borderWidth = 1;
        _consoleTextView.editable = NO;
        _consoleTextView.layoutManager.allowsNonContiguousLayout = NO;
        _consoleTextView.backgroundColor = HEXCOLOR(0xededed);
    }
    return _consoleTextView;
}

- (UIButton *)scanButton {
    if (!_scanButton) {
        _scanButton = [UIButton buttonWithType:UIButtonTypeCustom];
        _scanButton.layer.cornerRadius = 5;
        [_scanButton setTitle:@"Scan ble device" forState:UIControlStateNormal];
        _scanButton.backgroundColor = UIColor.orangeColor;
        [_scanButton addTarget:self action:@selector(scanMeshDevice) forControlEvents:UIControlEventTouchUpInside];
    }
    return _scanButton;
}


- (UIButton *)stopScanButton {
    if (!_stopScanButton) {
        _stopScanButton = [UIButton buttonWithType:UIButtonTypeCustom];
        _stopScanButton.layer.cornerRadius = 5;
        [_stopScanButton setTitle:@"Stop scan ble device" forState:UIControlStateNormal];
        _stopScanButton.backgroundColor = UIColor.orangeColor;
        [_stopScanButton addTarget:self action:@selector(stopScanMeshDeviceClick) forControlEvents:UIControlEventTouchUpInside];
    }
    return _stopScanButton;
}

- (NSMutableArray *)meshModelArray {
    if (!_meshModelArray) {
        _meshModelArray = [[NSMutableArray alloc] init];
    }
    return _meshModelArray;
}

- (NSMutableArray *)nodeIdArray {
    if (!_nodeIdArray) {
        _nodeIdArray = [[NSMutableArray alloc] init];
    }
    return _nodeIdArray;
}

- (NSMutableArray *)devIdArray {
    if (!_devIdArray) {
        _devIdArray = [[NSMutableArray alloc] init];
    }
    return _devIdArray;
}

@end

@interface TYAlertMessgeView ()
@property (nonatomic, strong) UIView *contentView;
@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UILabel *messageLabel;
@property (nonatomic, strong) CALayer *lineLayer;
@property (nonatomic, strong) UIButton *sureButton;
@property (nonatomic, strong) UIButton *cancelButton;
@property (nonatomic, copy) TYMessageHandlerBlock sureBtnClickBlock;
@property (nonatomic, copy) TYMessageHandlerBlock cancelBtnClickBlock;


@end

@implementation TYAlertMessgeView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self createView];
    }
    return self;
}


- (void)updateMessage:(NSString *)message {
    self.messageLabel.text = message;
    self.hidden = NO;
}

- (void)showTtile:(NSString *)title Message:(NSString *)message
    completeBlock:(TYMessageHandlerBlock)completeBlock
      cancelBlock:(TYMessageHandlerBlock)cancelBlock {
    self.titleLabel.text = title;
    self.messageLabel.text = message;
    self.sureBtnClickBlock = completeBlock;
    self.cancelBtnClickBlock = cancelBlock;
    self.hidden = NO;
}

- (void)createView {
    self.layer.cornerRadius = 5;
    self.layer.masksToBounds = YES;
    self.backgroundColor = HEXCOLOR(0xededed);
    [self addSubview:self.contentView];
    [self.contentView addSubview:self.titleLabel];
    [self.contentView addSubview:self.messageLabel];
    [self.contentView addSubview:self.sureButton];
    [self.contentView addSubview:self.cancelButton];
    [self.contentView.layer addSublayer:self.lineLayer];
    self.lineLayer.frame = CGRectMake(0, 50, APP_SCREEN_WIDTH, 1);
    self.contentView.frame = CGRectMake(0, 0, APP_SCREEN_WIDTH, 200);
    self.titleLabel.frame = CGRectMake(20, 20, APP_SCREEN_WIDTH - 40, 20);
    self.messageLabel.frame = CGRectMake(20, 60, APP_SCREEN_WIDTH - 40, 40);
    self.sureButton.frame = CGRectMake(20, 130, 150, 40);
    self.cancelButton.frame = CGRectMake(APP_SCREEN_WIDTH - 20 -150, 130, 150, 40);
}

- (void)sureButtonClick:(id)sender {
    if (self.sureBtnClickBlock) {
        self.sureBtnClickBlock();
    }
    self.hidden = YES;
}

- (void)cancelButtonClick:(id)sender {
    if (self.cancelBtnClickBlock) {
        self.cancelBtnClickBlock();
    }
    self.hidden = YES;
}

#pragma mark -- setter and getter methods

- (UIView *)contentView {
    if (!_contentView) {
        _contentView = [[UIView alloc] init];
    }
    return _contentView;
}

- (UILabel *)messageLabel {
    if (!_messageLabel) {
        _messageLabel = [[UILabel alloc] init];
        _messageLabel.font = [UIFont systemFontOfSize:15];
        _messageLabel.numberOfLines = 0;
        _messageLabel.textAlignment = NSTextAlignmentCenter;
    }
    return _messageLabel;
}

- (UIButton *)sureButton {
    if (!_sureButton) {
        _sureButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [_sureButton setTitle:@"去配网" forState:UIControlStateNormal];
        [_sureButton addTarget:self action:@selector(sureButtonClick:) forControlEvents:UIControlEventTouchUpInside];
        _sureButton.backgroundColor = UIColor.orangeColor;
    }
    return _sureButton;
}

- (UIButton *)cancelButton {
    if (!_cancelButton) {
        _cancelButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [_cancelButton setTitle:@"取消" forState:UIControlStateNormal];
        [_cancelButton addTarget:self action:@selector(cancelButtonClick:) forControlEvents:UIControlEventTouchUpInside];
        _cancelButton.backgroundColor = UIColor.orangeColor;
    }
    return _cancelButton;
}

- (UILabel *)titleLabel {
    if (!_titleLabel) {
        _titleLabel = [[UILabel alloc] init];
        _titleLabel.font =  [UIFont systemFontOfSize:15];
        _titleLabel.textAlignment = NSTextAlignmentCenter;
    }
    return _titleLabel;
}

- (CALayer *)lineLayer {
    if (!_lineLayer) {
        _lineLayer = [[CALayer alloc] init];
        _lineLayer.backgroundColor = [UIColor grayColor].CGColor;
    }
    return _lineLayer;
}

@end


TuyaSmartDemo/Classes/Login/Views/TYDemoLoginAndRegisterView.m
//
//  TYLoginAndRegisterView.m
//  TuyaSmartHomeKit_Example
//
//  Created by Kennaki Kai on 2018/11/30.
//  Copyright © 2018 xuchengcheng. All rights reserved.
//

#import "TYDemoLoginAndRegisterView.h"

@implementation TYDemoLoginAndRegisterView
{
    TYLoginAndRegisterViewType currentType;
}

- (instancetype)initWithType:(TYLoginAndRegisterViewType)type{
    if (self = [super init]) {
        currentType = type;
        [self initView];
        CGFloat top = IPhoneX ? 130 : 100;
        self.frame = CGRectMake(0, top, APP_SCREEN_WIDTH, APP_SCREEN_HEIGHT - 100);
    }
    return self;
}

- (void)initView {

    self.userInteractionEnabled = YES;
    UILabel *countryKeyLabel = [self keyLabel];
    countryKeyLabel.text = TYSDKDemoLocalizedString(@"Country code", @"");
    countryKeyLabel.frame = CGRectMake(20, 0, 120, 20);
    [self addSubview:countryKeyLabel];
    self.countryCodeField = [self valueTextField];
    self.countryCodeField.centerY = countryKeyLabel.centerY;
    [self addSubview:self.countryCodeField];
    self.countryCodeField.text = @"86";
    
    UILabel *accountKeyLabel = [self keyLabel];
    accountKeyLabel.text = TYSDKDemoLocalizedString(@"Account", @"");
    accountKeyLabel.frame = CGRectMake(20, CGRectGetMaxY(self.countryCodeField.frame) + 20 + 15, 120, 20);
    [self addSubview:accountKeyLabel];
    self.accountField = [self valueTextField];
    self.accountField.centerY = accountKeyLabel.centerY;
    [self addSubview:self.accountField];
    self.accountField.placeholder = TYSDKDemoLocalizedString(@"Phone number/Email", @"");
    
    UILabel *pwdKeyLabel = [self keyLabel];
    pwdKeyLabel.text = TYSDKDemoLocalizedString(@"input_password", @"");
    pwdKeyLabel.frame = CGRectMake(20, CGRectGetMaxY(self.accountField.frame) + 20 + 15, 120, 20);
    [self addSubview:pwdKeyLabel];
    self.passwordField = [self valueTextField];
    self.passwordField.centerY = pwdKeyLabel.centerY;
    [self addSubview:self.passwordField];
    self.passwordField.placeholder = TYSDKDemoLocalizedString(@"input_password", @"");
    self.passwordField.secureTextEntry = YES;
    if (@available(iOS 12.0, *)) {
        self.passwordField.textContentType = UITextContentTypeOneTimeCode;
    } else {
        
    }
    
    switch (currentType) {
        case TYLoginAndRegisterViewTypeLogin:
        {
            UIButton *pwdLoginButton = [UIButton buttonWithType:UIButtonTypeCustom];
            [pwdLoginButton setTitle:TYSDKDemoLocalizedString(@"Login with password", @"") forState:UIControlStateNormal];
            pwdLoginButton.frame = CGRectMake(20, CGRectGetMaxY(self.passwordField.frame) + 20, APP_SCREEN_WIDTH - 40, 44);
            pwdLoginButton.layer.cornerRadius = 5;
            pwdLoginButton.backgroundColor = UIColor.orangeColor;
            [pwdLoginButton addTarget:self action:@selector(passwordLogin) forControlEvents:UIControlEventTouchUpInside];
            [self addSubview:pwdLoginButton];
        }
            break;
        case TYLoginAndRegisterViewTypeRegister:
        {
            UILabel *verifyCodeKeyLabel = [self keyLabel];
            verifyCodeKeyLabel.text = TYSDKDemoLocalizedString(@"Verify code", @"");
            verifyCodeKeyLabel.frame = CGRectMake(20, CGRectGetMaxY(self.passwordField.frame) + 20 + 15, 120, 20);
            [self addSubview:verifyCodeKeyLabel];
            self.verifyCodeField = [self valueTextField];
            self.verifyCodeField.centerY = verifyCodeKeyLabel.centerY;
            [self addSubview:self.verifyCodeField];
            self.verifyCodeField.placeholder = TYSDKDemoLocalizedString(@"Verify code", @"");
            
            CGFloat buttonWidth = (APP_SCREEN_WIDTH - 60)/2;
            
            UIButton *sendVerifyCodeButton = [UIButton buttonWithType:UIButtonTypeCustom];
            sendVerifyCodeButton.backgroundColor = UIColor.orangeColor;
            [sendVerifyCodeButton setTitle:TYSDKDemoLocalizedString(@"Send code", @"") forState:UIControlStateNormal];
            sendVerifyCodeButton.frame = CGRectMake(20, CGRectGetMaxY(self.verifyCodeField.frame) + 20, buttonWidth, 44);
            sendVerifyCodeButton.layer.cornerRadius = 5;
            [sendVerifyCodeButton addTarget:self action:@selector(sendVerifyCode) forControlEvents:UIControlEventTouchUpInside];
            [self addSubview:sendVerifyCodeButton];
            
            UIButton *registerButton = [UIButton buttonWithType:UIButtonTypeCustom];
            registerButton.backgroundColor = UIColor.orangeColor;
            [registerButton setTitle:TYSDKDemoLocalizedString(@"login_register", @"") forState:UIControlStateNormal];
            registerButton.frame = CGRectMake(buttonWidth + 40/*button space 20 and left space 20*/, CGRectGetMaxY(self.verifyCodeField.frame) + 20, buttonWidth, 44);
            registerButton.layer.cornerRadius = 5;
            [registerButton addTarget:self action:@selector(registerAction) forControlEvents:UIControlEventTouchUpInside];
            [self addSubview:registerButton];
        }
            break;
        default:
            break;
    }
   
    UITapGestureRecognizer *tapToDismiss = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(dismissKeyboard)];
    [self addGestureRecognizer:tapToDismiss];
}

- (void)passwordLogin {
    if (self.delegate && [self.delegate respondsToSelector:@selector(loginAndRegisterViewTrigerredAction:)]) {
        [self.delegate loginAndRegisterViewTrigerredAction:TYLoginAndRegisterViewActionTypeLogin];
    }
}

- (void)sendVerifyCode {
    if (self.delegate && [self.delegate respondsToSelector:@selector(loginAndRegisterViewTrigerredAction:)]) {
        [self.delegate loginAndRegisterViewTrigerredAction:TYLoginAndRegisterViewActionTypeSendVerifyCode];
    }
}

- (void)registerAction {
    if (self.delegate && [self.delegate respondsToSelector:@selector(loginAndRegisterViewTrigerredAction:)]) {
        [self.delegate loginAndRegisterViewTrigerredAction:TYLoginAndRegisterViewActionTypeRegister];
    }
}

- (void)dismissKeyboard {
    [self endEditing:YES];
    [self.countryCodeField resignFirstResponder];
    [self.accountField resignFirstResponder];
    [self.passwordField resignFirstResponder];
    [self.verifyCodeField resignFirstResponder];
}

- (UILabel *)keyLabel {
    UILabel *label = [UILabel new];
    label.font = [UIFont systemFontOfSize:15];
    return label;
}

- (UITextField *)valueTextField {
    UITextField *textField = [[UITextField alloc] initWithFrame:CGRectMake(140, 0, APP_SCREEN_WIDTH - 140 - 20, 44)];
    textField.layer.borderColor = UIColor.blackColor.CGColor;
    textField.layer.borderWidth = 0.5;
    textField.layer.cornerRadius = 5;
    textField.layer.masksToBounds = YES;
    textField.delegate = self;
    return textField;
}

- (UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event {
    UIView *view = [super hitTest:point withEvent:event];
    if (!view) {
        for (UIView *aView in self.subviews) {
            CGPoint myPoint = [aView convertPoint:point fromView:self];
            if (CGRectContainsPoint(aView.bounds, myPoint)) {
                return aView;
            }
        }
    }
    return view;
}

#pragma mark - UITextFieldDelegate delegate

- (void)textFieldDidBeginEditing:(UITextField *)textField {
    
    if (textField == self.verifyCodeField) {
        self.top -= 150;
    }
}

- (void)textFieldDidEndEditing:(UITextField *)textField {
    
    CGFloat top = IPhoneX ? 130 : 100;
    self.top = top;
}
@end
